Add support for setting/getting Date objects

sbencoding opened the pull request:
Implementing a new feature described in issue https://github.com/sindresorhus/electron-store/issues/18.  
Type information is stored under `${INTERNAL_KEY}.type` as suggested, the `get` method checks if there's a type key present, if there's it converts the value to the type before returning it.  
The `set` methods checks if the value passed in is an `instance of Date`, if it is it stores it as an object like:
```js
key: {
    __internal__.type: 'Date',
    val: valuePassedIn
}
````
I've also added a test to `test.js` to test for the setting and getting Date objects
Let me know if there's anything I could change. For example I chose to include type information only for Date objects, but it might be better to store type information for every key-value pair to be more consistent, but it would also make the object and the file larger.

sbencoding commented:
The travis CI tests/builds are passing, but node v12 on Windows failed to download node.js v12, so the code should be fine I think.

sindresorhus commented:
This needs to be documented in the readme and index.d.ts. Also make sure that `$ npm test` passes locally.

sindresorhus commented on index.d.ts:
```diff
@@ -202,6 +202,13 @@ declare namespace Conf {
 		readonly accessPropertiesByDotNotation?: boolean;
 	}
+
+	interface ExtraType<T> {
```
The interface and the individual properties need doc comments.

sindresorhus commented on index.d.ts:
```diff
@@ -202,6 +202,13 @@ declare namespace Conf {
+	interface ExtraType<T> {
+		name: string,
+		instance: T,
+		convertFrom: (val: T) => string | number,
```
The return value here would be a good candidate for https://github.com/sindresorhus/type-fest/blob/eb110db80ccd74e8f5141cb52b3e99d3603399ce/source/basic.d.ts#L53

sindresorhus commented on index.d.ts:
```diff
@@ -248,7 +256,7 @@ declare class Conf<T = any> implements Iterable<[keyof T, T[keyof T]]> {
 
 	@param key - You can use [dot-notation](https://github.com/sindresorhus/dot-prop) in a key to access nested properties.
-	@param value - Must be JSON serializable. Trying to set the type `undefined`, `function`, or `symbol` will result in a `TypeError`.
+	@param value - Must be JSON serializable. Trying to set the type `undefined`, `function`, or `symbol` will result in a `TypeError`. Additionally Conf automatically seralizes and deserializes `Date` objects
```
Typo

sindresorhus commented on index.d.ts:
```diff
@@ -211,6 +218,7 @@ declare class Conf<T = any> implements Iterable<[keyof T, T[keyof T]]> {
 	store: T;
 	readonly path: string;
 	readonly size: number;
+	readonly extraTypes: Conf.ExtraType<any>[];
```
Should not be `any`.

sindresorhus commented on index.js:
```diff
@@ -214,6 +224,75 @@ class Conf {
+		// Sub-object handling of extra types
+		const stack = [];
+		stack.push(value);
+		while (stack.length > 0) {
```
Why are you using a while loop?

sbencoding replied:
So in case of nested objects I want to reuse the same logic to serialize `Date` types, so there are two ways I can go about this. The first one is to use recursion, the other one is to use the iterative solution, which is the while loop.  
I've decided to use the iterative solution over the recursive solution, because with really deeply nested objects, there's a possibility that the function could overflow the callstack.

sindresorhus commented on readme.md:
```diff
@@ -321,6 +323,14 @@ conf.store = {
+Currently supported objects are:
+- `Date` - serializes the Date to milliseconds, then back to a `Date` object when the getter is called
+
+If you want support for a new object/type please open a new issue for discussion.  
```
The docs in the readme and index.d.ts should be in sync.

sbencoding commented on test.js:
```diff
@@ -47,6 +47,18 @@ test('.set() - with object', t => {
+	t.context.config.set('dateobj', {str: 'this is a string', date: new Date()});
+	const myDate = t.context.config.get('dateobj').date;
+	t.true(myDate instanceof Date);
+});
```
I've added more tests, but I'm not sure how to go about circular objects, becuase `Conf.serialize`, which is `JSON.stringify` by default, throws an exception is there's a circulation. So all I did for now is to detect if there's a circulation and stop handling that branch of the object, because it would cause an infinite loop, and cause the application to crash, but when `Conf` gets to serializing the object with `JSON` it still throws an error. I've also tested this behavior with `string` so it's not a problem with how the new feature handles extra types.

delewis13 commented:
Any updates on merging this functionality? Would be great to have a simple way to store and retrieve dates.

sbencoding commented:
@delewis13 I've implemented the requested changes, but I didn't want to update the PR, because there are still questions, namely:
* Circular objects
In the review there's a request to handle circular objects, but even if I handle them `JSON.serialize()` just throws and exception when it detects one, so I'm not sure what to do here
* Usage of while loops
I've commented on the review explaining my thought process, but there might be other, better ways instead of using the while loop I haven't thought about.

Other than that the rest of the requested changes are implemented

papb commented:
Hi @sbencoding, there are other modules that can handle serialization and deserialization of circular objects automatically for you, such as [flatted](https://github.com/WebReflection/flatted), maybe this is what Sindre had in mind...

sbencoding commented:
Hey, sorry for the late response.
I've implemented the serialization and deserialization of `circular objects`, as well as I've added some tests for it. Other requested changes are implemented as well.
I'm not sure why the travis build fails tho....
* On all windows builds `xo` complains about import order in `test.js`, this doesn't happen for OSX or Linux or my local machine (Linux).
* OSX with nodejs v12 fails the both of the `watch` option tests. This doesn't happen for v10, and on the linux tests.

sindresorhus commented on index.d.ts:
```diff
@@ -1,9 +1,30 @@
 /// <reference types="node"/>
-import {JSONSchema} from 'json-schema-typed';
+import { JSONSchema } from 'json-schema-typed';
```
Don't do unrelated style changes. Applies in many places. If you upgrade the XO dependency and run `npx xo --fix` these things will be fixed for you.

sindresorhus commented on index.d.ts:
```diff
@@ -1,9 +1,30 @@
 declare namespace Conf {
 	type Schema = JSONSchema;
 
+	/**
+	 * Matches a JSON object
+	 */
```
```suggestion
	/**
	Matches a JSON object.
	*/
```

Applies everywhere.

sindresorhus commented on index.d.ts:
```diff
@@ -225,6 +268,10 @@ declare class Conf<T = any> implements Iterable<[keyof T, T[keyof T]]> {
 	readonly path: string;
 	readonly size: number;
+	/**
+	 * Conf can automatically (de)serialize some objects/types, for more information please visit [Supported Types](https://https://github.com/sindresorhus/conf#Supported-Types)
```
The readme and index.d.ts docs should be in sync. That means the linked to text should be included inline here.

sindresorhus commented on index.js:
```diff
@@ -283,6 +294,126 @@ class Conf {
+		let circularReferenceCounter = 1;
+		while (stack.length > 0) {
+			const {current, parent, key} = stack.pop();
+			if (objectSeenList.indexOf(current) === -1) {
```
```suggestion
			if (!objectSeenList.includes(current)) {
```

sindresorhus commented on index.js:
```diff
@@ -283,6 +294,126 @@ class Conf {
+		}
+
+		return value;
+	}
```
This method is pretty large. I would look into splitting it into multiple methods and maybe something here is generic enough to be moved out into a generic utility function. Same with `_deserializeExtraTypes`.

sindresorhus commented on index.js:
```diff
@@ -283,6 +294,126 @@ class Conf {
+				// seenList.indexOf(current) is the object it's pointing to
+				const referenceTarget = objectSeenList[objectSeenList.indexOf(current)];
+				parent[key] = {
+					[TYPE_KEY]: 'circularObject',
```
I don't really see why you need the `circularObject` thing. Circular objects can simply be detected by comparing references.  I think this logic could be simplified a lot.

See how it's done here: https://github.com/sindresorhus/serialize-error/blob/master/index.js and here: https://github.com/sindresorhus/map-obj/blob/master/index.js

sbencoding replied:
The `circularObject` type signals to the deserialize function that it was a circular object, so it can lookup the `referenceNumber` set in the value of the object, and set it to the object it was referencing to before serialization.

I might have misunderstood this problem. I made it so that when deserializing the circulation is reconstructed, so you're going to end up with the same circular object after getting it from the store.
So should I just completely remove circulation like the `destroyCircular` function does in the first example you've provided, or is there an easier way to reconstruct the circulation?

sindresorhus commented on test.js:
```diff
@@ -49,6 +49,76 @@ test('.set() - with object', t => {
+	t.true(getCircular.c1.myDate instanceof Date);
+	t.true(getCircular.c2.myDate instanceof Date);
+});
+
```
There needs to be some tests for other types, just to ensure it works in general with custom types and other types than Date.

sindresorhus commented on index.js:
```diff
@@ -283,6 +294,126 @@ class Conf {
+					console.log(referenceTarget[CIRCULAR_REF_KEY]);
+				}
+
+				console.log('skipping pontentially circulation');
```
Modules should not log to the console.

sbencoding replied:
Yeah, I'm very, very sorry about this, sometimes I just add some log statements for testing, and I always try to keep in mind to remove them, but it seems like I've forgotten this time.

sindresorhus commented on readme.md:
```diff
@@ -356,6 +358,15 @@ conf.store = {
 
```
You need to document how to use the custom deserializer/serializer.

sbencoding commented:
I've implemented most of the requested changes, I'm just going to re-work a bit the addition of extra types by the consumer program tomorrow, and then I'm going push the new changes.

I'm really confused about what I should do with circular objects now, but that's explained in a comment in the code review. I'd be happy to implement another approach, if I'm missing the point of this feature.

sbencoding commented:
I've implemented the requested changes, now the build only fails for node v12 on macOS during the 2 `.watch()` tests.

Instead of directly updating the extraTypes array of a Conf instance, I chose to implement the addition of extra types through the options argument that's passed to the constructor.

Let me know if there's anything else to change.

sindresorhus commented on readme.md:
```diff
@@ -356,6 +358,77 @@ conf.store = {
+- It must return the type that is defined.
+
+#### Example
+The following example demonstrates the configuration of the `Date` type, that is internally supported by default.
```
It would be better to use a different type than the default.

sindresorhus commented:
There are still some unresolved inline comments. Also make sure that the docs in the readme is in sync with index.d.ts and the other way.

sbencoding commented:
I've updated the example and double-checked to docs to reflect each other as much as possible, let me know if there's anything else to change/update.

sindresorhus commented:
I appreciate all the work done here, but I unfortunately don't think this is the optimal solution. Closing in favor of #154.
